### PR TITLE
Introduce WaitCursorScope

### PR DESCRIPTION
**Pull request by drewnoakes**
`WaitCursorScope` allows setting the current mouse cursor as 'busy' within a given scope, and having it automatically restore at the end of that scope.

```c#
using (new WaitCursorScope())
{
    // ...
}
```

### Pros:

- Nests properly. If one scope exists within another, exiting the enclosed scope doesn't reset the cursor back to 'normal'. Exiting a scope restores the cursor to whatever it was when the scope was entered. A nested scope sets the wait cursor on both entry and exit, which makes sense as the outer scope is 'busy' throughout.
- Exception safe. No matter what happens within the `using` block, the (global, singleton) cursor will be restored properly.
- Refactor safe. No need to be mindful of two unrelated commands located potentially some distance apart. Introducing a return within the block won't leave your mouse busy.

### Cons:

- Code within the block must be indented.
- Requires scope to exist within one method. I found two cases where this didn't hold: `FormResolveConflicts` and `FormStash`, but the two approaches can co-exist just fine.

What did I do to test the code and ensure quality:
 - Manual testing
 - Manual review (changes are very simple -- [ignore white-space in diff](https://github.com/gitextensions/gitextensions/pull/4670/files?w=1))

Has been tested on:
 - GIT 2.16
 - Windows 10

### CODE REVIEW

File: GitUI/WaitCursorScope.cs [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    /// }
+    /// </code>
+    /// This pattern ensures:
+    /// <list type="bullet">
+    /// <item>The cursor is always restored when the block exits, whether by fall-through, return or exception.</item>
+    /// <item>Nested scopes work properly. That is, an inner scope doesn't reset the mouse when it exits.
+    /// Only the top-most scope will fully restore the mouse cursor.</item>
+    /// </list>
+    /// </remarks>
+    internal sealed class WaitCursorScope : IDisposable
```
Comment (sharwell) [inline GitUI/WaitCursorScope.cs]: :bulb: `struct`

Comment (sharwell) [reply GitUI/WaitCursorScope.cs]: :memo: you'll need to replace the parameterless constructor with a factory method, but it shouldn't be a problem.

Comment (drewnoakes) [reply GitUI/WaitCursorScope.cs]: I'd considered using a struct but assumed the value would be boxed for the IDisposable call. Maybe the compiler binds directly? Any other reason this should be a struct?

**Comment by RussKie** [comment]
Eric Lippert said it is possible and no boxing occur
https://stackoverflow.com/a/2413844
And i agree it looks odd...

**Comment by drewnoakes** [comment]
Indeed no boxing occurs.

```c#
class Test
{
    void Foo()
    {
        using (new DisposeMe()) {}
    }
}

struct DisposeMe : IDisposable
{
    public void Dispose() { }
}
```

The MSIL for `Foo`:

```
  .method private hidebysig instance void 
    Foo() cil managed 
  {
    .maxstack 1
    .locals init (
      [0] valuetype DisposableStructTest.DisposeMe V_0
    )

    IL_0000: ldloca.s     V_0
    IL_0002: initobj      DisposableStructTest.DisposeMe
    .try
    {
      IL_0008: leave.s      IL_0018
    }
    finally
    {
      IL_000a: ldloca.s     V_0
      IL_000c: constrained. DisposableStructTest.DisposeMe
      IL_0012: callvirt     instance void [netstandard]System.IDisposable::Dispose()
      IL_0017: endfinally   
    }
    IL_0018: ret          
  }
```

Use of [`OpCodes.Constrained`](https://docs.microsoft.com/en-us/dotnet/api/system.reflection.emit.opcodes.constrained?view=netframework-4.7.1) at `IL_000c`. No boxing.

Interesting that the compiler uses `callvirt` here, when the exact type and method are known. `callvirt` is sometimes used over `call` even when devirtualisation could be a applied to calls of virtual members of reference types, because `callvirt` does a null check. The value cannot be null here. Maybe there is another reason. Not a biggie, just a curiosity.

I can change this over to a `struct` later today.

**Comment by drewnoakes** [comment]
Rebased, fixed conflicts, did more testing.

@RussKie here's another example where I wonder what you'd want me to do. Since being marked as 'needs review' it's received two accepting reviews. Should I mark it as 'ready'? Should I just merge it?

**Comment by RussKie** [comment]
I want to pull down and check it first, if that's ok

**Comment by RussKie** [comment]
Awesome work, thank you

**Pull request merged**